File: src/application-context.ts
import { posix } from "node:path";
import { CycleCloudRequestError } from "./errors.js";
import type { ImagePlatform } from "./image-platform.js";
import {
    compareNames,
    consumedFields,
    requiredRecord,
    requiredWireString,
} from "./normalize.js";

export type ApplicationContextSection =
    | "environment"
    | "storage"
    | "attachments";
export type ApplicationReadSelection =
    | { view: "overview" }
    | {
          view: "details";
          targetName: string;
          section: ApplicationContextSection;
      };
export interface ContextPage {
    readonly offset?: number;
    readonly limit?: number;
}
export interface ApplicationContextInput {
    readonly clusterName: string;
    readonly targetNames?: readonly string[];
    readonly installPath: string;
    readonly targetLimit: number;
    readonly view?: "overview" | "details";
    readonly section?: ApplicationContextSection;
    readonly offset?: number;
    readonly itemLimit?: number;
}

export type Unavailable = { available: false; warning: string };
export interface AvailableList<T> {
    available: true;
    items: T[];
    total: number;
    returned: number;
    truncated: boolean;
    offset: number;
    nextOffset: number | null;
}
export type EvidenceList<T> = AvailableList<T> | Unavailable;
type Fields = ReadonlyMap<string, unknown>;

export function unavailable(warning: string): Unavailable {
    return { available: false, warning };
}

export function boundedList<T>(
    items: readonly T[],
    limit = 20,
    offset = 0,
    byteLimit = 6144,
): AvailableList<T> {
    const page: T[] = [];
    let bytes = 2;
    for (const item of items.slice(offset, offset + limit)) {
        const size = Buffer.byteLength(JSON.stringify(item), "utf8") + 1;
        if (bytes + size > byteLimit) {
            if (!page.length) invalid();
            break;
        }
        page.push(item);
        bytes += size;
    }
    return {
        available: true,
        items: page,
        total: items.length,
        returned: page.length,
        truncated: items.length > page.length,
        offset,
        nextOffset:
            offset + page.length < items.length ? offset + page.length : null,
    };
}

function invalid(): never {
    throw new CycleCloudRequestError("invalid_response", false);
}
function text(value: unknown): string | undefined {
    return value == null || value === ""
        ? undefined
        : requiredWireString(value, 256);
}
function flag(value: unknown): boolean | undefined {
    if (value == null) return undefined;
    if (typeof value !== "boolean") invalid();
    return value;
}
function integer(value: unknown): number | undefined {
    if (value == null) return undefined;
    if (typeof value !== "number" || !Number.isSafeInteger(value)) invalid();
    return value;
}
function identifier(value: unknown): string | undefined {
    const result = text(value);
    // These fields identify configured resources, not credential-bearing URLs.
    if (result !== undefined && /:\/\/|[?#]/u.test(result)) invalid();
    return result;
}
function rows(value: unknown): unknown[] {
    if (!Array.isArray(value)) invalid();
    return value;
}
function unique<T>(items: T[], name: (item: T) => string): T[] {
    const keys = items.map(name).map((key) => key.toLowerCase());
    if (new Set(keys).size !== keys.length) invalid();
    return items.sort((left, right) => compareNames(name(left), name(right)));
}
function recordList<T>(
    value: unknown,
    normalize: (name: string, fields: unknown) => T,
    page: ContextPage = {},
): EvidenceList<T> {
    if (value == null)
        return unavailable(
            "This configuration collection was not supplied by CycleCloud.",
        );
    const entries = Object.entries(requiredRecord(value));
    for (const [name] of entries) requiredWireString(name, 256);
    const items = unique(entries, ([name]) => name).map(([name, fields]) =>
        normalize(name, fields),
    );
    return boundedList(items, page.limit, page.offset);
}

export function isInstallPath(value: string): boolean {
    try {
        requiredWireString(value, 1024);
    } catch {
        return false;
    }
    return (
        value.startsWith("/") &&
        !value.startsWith("//") &&
        !value.split("/").some((part) => part === "." || part === "..")
    );
}

function stringList(
    value: unknown,
    page: ContextPage = {},
): EvidenceList<string> {
    if (value == null)
        return unavailable(
            "This configuration field was not supplied by CycleCloud.",
        );
    const items = typeof value === "string" ? [value] : rows(value);
    const strings = items.map((item) => requiredWireString(item, 256));
    if (
        new Set(strings.map((item) => item.toLowerCase())).size !==
        strings.length
    )
        invalid();
    // Inheritance precedence and VM preference order are configuration evidence.
    return boundedList(strings, page.limit, page.offset);
}

function normalizeSpecs(value: unknown, page: ContextPage = {}) {
    return recordList(
        value,
        (_name, raw) => {
            const f = consumedFields(raw, [
                "project",
                "spec",
                "version",
                "sourcelocker",
                "additionalspec",
                "order",
            ]);
            return {
                project: requiredWireString(f.get("project"), 256),
                spec: requiredWireString(f.get("spec"), 256),
                version: identifier(f.get("version")),
                sourceLocker: identifier(f.get("sourcelocker")),
                additional: flag(f.get("additionalspec")),
                order: integer(f.get("order")),
            };
        },
        page,
    );
}

function normalizeMounts(
    value: unknown,
    installPath: string,
    page: ContextPage = {},
) {
    return recordList(
        value,
        (name, raw) => {
            const f = consumedFields(raw, [
                "mountpoint",
                "type",
                "fs_type",
                "disabled",
            ]);
            const mountpoint = text(f.get("mountpoint"));
            if (mountpoint !== undefined && !isInstallPath(mountpoint))
                invalid();
            const disabled = flag(f.get("disabled"));
            const prefix =
                mountpoint === undefined
                    ? undefined
                    : posix.normalize(mountpoint).replace(/\/$/u, "");
            const within =
                prefix !== undefined &&
                (installPath === prefix ||
                    installPath.startsWith(`${prefix}/`));
            return {
                name,
                mountpoint,
                type: text(f.get("type")),
                filesystem: text(f.get("fs_type")),
                disabled,
                coversInstallPath:
                    disabled === true || (prefix !== undefined && !within)
                        ? false
                        : disabled === false && prefix !== undefined
                          ? within
                          : null,
            };
        },
        page,
    );
}

function normalizeVolumes(value: unknown, page: ContextPage = {}) {
    return recordList(
        value,
        (name, raw) => {
            const f = consumedFields(raw, ["mount", "persistent", "disabled"]);
            return {
                name,
                mount: text(f.get("mount")),
                persistent: flag(f.get("persistent")),
                disabled: flag(f.get("disabled")),
            };
        },
        page,
    );
}

function normalizeAttachment(
    value: unknown,
): { available: true; parameterName: string } | Unavailable {
    if (typeof value !== "string")
        return unavailable(
            "No simple additional-spec parameter reference is available; inspect the cluster template before attaching.",
        );
    const match =
        /^\$(?:([A-Za-z_][A-Za-z0-9_.-]{0,255})|\{([A-Za-z_][A-Za-z0-9_.-]{0,255})\})$/u.exec(
            value,
        );
    const name = match?.[1] ?? match?.[2];
    return name === undefined
        ? unavailable(
              "The attachment reference is not a simple parameter; inspect the cluster template before attaching.",
          )
        : { available: true, parameterName: name };
}

function normalizeIdentity(f: Fields) {
    const name = requiredWireString(f.get("name"), 256);
    const array = flag(f.get("isarray"));
    if (array !== true && text(f.get("template")) !== name) invalid();
    return {
        name,
        kind: array === true ? ("nodearray" as const) : ("node" as const),
    };
}

export function normalizeApplicationOverview(value: unknown) {
    return unique(
        rows(value).map((raw) => {
            const f = consumedFields(raw, [
                "name",
                "template",
                "isarray",
                "state",
                "imagename",
                "slurmrole",
                "slurmpartition",
                "slurmhaenabled",
                "attachmentreference",
            ]);
            const identity = normalizeIdentity(f);
            const attachment = normalizeAttachment(
                f.get("attachmentreference"),
            );
            return {
                ...identity,
                state: text(f.get("state")),
                image: identifier(f.get("imagename")),
                role: text(f.get("slurmrole")),
                partition: text(f.get("slurmpartition")),
                haEnabled: flag(f.get("slurmhaenabled")),
                attachmentParameter: attachment.available
                    ? attachment.parameterName
                    : undefined,
            };
        }),
        (node) => node.name,
    );
}

export function normalizeApplicationDetails(
    value: unknown,
    installPath: string,
    section: ApplicationContextSection,
    page: ContextPage,
) {
    const sectionFields = {
        environment: [
            "extends",
            "state",
            "targetstate",
            "imagename",
            "machinetype",
            "architecture",
            "locker",
            "slurmrole",
            "slurmversion",
            "slurmpartition",
            "slurmhaenabled",
            "slurmprimaryscheduler",
            "slurmautoscale",
        ],
        storage: ["mounts", "volumes"],
        attachments: ["clusterinitspecs", "attachmentreference"],
    };
    return unique(
        rows(value).map((raw) => {
            const f = consumedFields(raw, [
                "name",
                "template",
                "isarray",
                ...sectionFields[section],
            ]);
            const identity = normalizeIdentity(f);
            if (section === "storage")
                return {
                    ...identity,
                    mounts: normalizeMounts(f.get("mounts"), installPath, page),
                    volumes: normalizeVolumes(f.get("volumes"), page),
                };
            if (section === "attachments")
                return {
                    ...identity,
                    specs: normalizeSpecs(f.get("clusterinitspecs"), page),
                    attachment: normalizeAttachment(
                        f.get("attachmentreference"),
                    ),
                };
            const image = identifier(f.get("imagename"));
            return {
                ...identity,
                state: text(f.get("state")),
                targetState: text(f.get("targetstate")),
                bases: stringList(f.get("extends"), page),
                image,
                platform: unavailable(
                    image === undefined
                        ? "The configured image was not supplied; its OS release is unresolved."
                        : "Image platform metadata could not be retrieved or validated.",
                ) as ImagePlatform,
                machineTypes: stringList(f.get("machinetype"), page),
                architecture: text(f.get("architecture")),
                locker: identifier(f.get("locker")),
                scheduler: {
                    role: text(f.get("slurmrole")),
                    version: text(f.get("slurmversion")),
                    partition: text(f.get("slurmpartition")),
                    haEnabled: flag(f.get("slurmhaenabled")),
                    primary: flag(f.get("slurmprimaryscheduler")),
                    autoscale: flag(f.get("slurmautoscale")),
                },
            };
        }),
        (node) => node.name,
    );
}

export type ApplicationOverview = ReturnType<
    typeof normalizeApplicationOverview
>[number];
export type ApplicationDetail = ReturnType<
    typeof normalizeApplicationDetails
>[number];

export function normalizeApplicationParameters(
    value: unknown,
    page: ContextPage = {},
) {
    return unique(
        rows(value).map((raw) => {
            const f = consumedFields(raw, [
                "name",
                "label",
                "parametertype",
                "value",
            ]);
            if (f.get("parametertype") !== "Cloud.ClusterInitSpecs") invalid();
            return {
                name: requiredWireString(f.get("name"), 256),
                label: text(f.get("label")),
                specs: normalizeSpecs(f.get("value"), page),
            };
        }),
        (parameter) => parameter.name,
    );
}

export function normalizeContextCluster(value: unknown, name: string) {
    const items = rows(value);
    if (!items.length)
        throw new CycleCloudRequestError("cluster_not_found", false);
    if (items.length !== 1) invalid();
    const f: Fields = consumedFields(items[0], [
        "clustername",
        "parentname",
        "state",
        "targetstate",
    ]);
    if (requiredWireString(f.get("clustername"), 256) !== name) invalid();
    return {
        name,
        parentName: text(f.get("parentname")),
        state: text(f.get("state")),
        targetState: text(f.get("targetstate")),
    };
}

export type ApplicationParameter = ReturnType<
    typeof normalizeApplicationParameters
>[number];
export interface ScopedParameter extends ApplicationParameter {
    usedBy: EvidenceList<string>;
}
export interface ApplicationContextResult {
    context: {
        clusterName: string;
        state?: string;
        targetState?: string;
        observedAt: string;
        evidence: "configured";
        view: "overview" | "details";
        section?: ApplicationContextSection;
        nextStep: string;
        installPath: string;
        sources: string[];
        targets:
            | (AvailableList<ApplicationOverview | ApplicationDetail> & {
                  missingRequested: string[];
              })
            | Unavailable;
        attachmentParameters?:
            | (AvailableList<ScopedParameter> & {
                  clusterName: string;
                  missingReferenced: string[];
              })
            | Unavailable;
        warnings: string[];
        unverified: string[];
    };
}

File: python/cyclecloud_agent_inspect/application_context.py
"""Configured application evidence, bounded pages and section normalization."""
import copy
import posixpath
import re

from .contract import (
    MISSING, PAGE_BYTE_LIMIT, is_install_path, json_bytes, name_sort_key,
    omit_missing, required_record, required_rows, required_wire_string,
    safe_integer, validate_limit,
)
from .errors import cluster_not_found, invalid_response
from .normalize import consumed_fields


def unavailable(warning):
    return {"available": False, "warning": warning}


def bounded_list(items, limit=20, offset=0, byte_limit=PAGE_BYTE_LIMIT):
    """Page by actual entries; truncation compares the page with the full collection."""
    items = required_rows(items)
    limit = validate_limit(limit, 0, 9007199254740991)
    offset = validate_limit(offset, 0, 1000000)
    byte_limit = validate_limit(byte_limit, 2, 9007199254740991)
    page = []
    size = 2
    for item in items[offset:offset + limit]:
        entry_size = len(json_bytes(item)) + 1
        if size + entry_size > byte_limit:
            if not page:
                invalid_response()
            break
        # The standalone page helper also owns its result, not just normalizers.
        page.append(copy.deepcopy(item))
        size += entry_size
    return {"available": True, "items": page, "total": len(items), "returned": len(page),
            "truncated": len(items) > len(page), "offset": offset,
            "nextOffset": offset + len(page) if offset + len(page) < len(items) else None}


def _text(value):
    if value is MISSING or value is None or value == "":
        return MISSING
    return required_wire_string(value, 256)


def _flag(value):
    if value is MISSING or value is None:
        return MISSING
    if not isinstance(value, bool):
        invalid_response()
    return value


def _integer(value):
    return MISSING if value is MISSING or value is None else safe_integer(value)


def _identifier(value):
    result = _text(value)
    if result is not MISSING and re.search(r"://|[?#]", result):
        invalid_response()
    return result


def _unique(items, name):
    names = [name(item).lower() for item in items]
    if len(set(names)) != len(names):
        invalid_response()
    return sorted(items, key=lambda item: name_sort_key(name(item)))


def _record_list(value, normalize, page):
    if value is MISSING or value is None:
        return unavailable("This configuration collection was not supplied by CycleCloud.")
    entries = [(required_wire_string(name, 256), fields) for name, fields in required_record(value).items()]
    items = [normalize(name, fields) for name, fields in _unique(entries, lambda entry: entry[0])]
    return bounded_list(items, page.get("limit", 20), page.get("offset", 0))


def _string_list(value, page):
    if value is MISSING or value is None:
        return unavailable("This configuration field was not supplied by CycleCloud.")
    values = [value] if isinstance(value, str) else required_rows(value)
    strings = [required_wire_string(item, 256) for item in values]
    if len(set(item.lower() for item in strings)) != len(strings):
        invalid_response()
    # Preserve inheritance precedence and configured VM preference order.
    return bounded_list(strings, page.get("limit", 20), page.get("offset", 0))


def _specs(value, page):
    def normalize(name, raw):
        fields = consumed_fields(raw, ("project", "spec", "version", "sourcelocker", "additionalspec", "order"))
        return omit_missing({"project": required_wire_string(fields.get("project", MISSING), 256),
                             "spec": required_wire_string(fields.get("spec", MISSING), 256),
                             "version": _identifier(fields.get("version", MISSING)),
                             "sourceLocker": _identifier(fields.get("sourcelocker", MISSING)),
                             "additional": _flag(fields.get("additionalspec", MISSING)),
                             "order": _integer(fields.get("order", MISSING))})
    return _record_list(value, normalize, page)


def _mounts(value, install_path, page):
    def normalize(name, raw):
        fields = consumed_fields(raw, ("mountpoint", "type", "fs_type", "disabled"))
        mountpoint = _text(fields.get("mountpoint", MISSING))
        if mountpoint is not MISSING and not is_install_path(mountpoint):
            invalid_response()
        disabled = _flag(fields.get("disabled", MISSING))
        prefix = MISSING if mountpoint is MISSING else posixpath.normpath(mountpoint).rstrip("/")
        within = prefix is not MISSING and (install_path == prefix or install_path.startswith(prefix + "/"))
        if disabled is True or (prefix is not MISSING and not within):
            covers = False
        elif disabled is False and prefix is not MISSING:
            covers = within
        else:
            covers = None
        return omit_missing({"name": name, "mountpoint": mountpoint, "type": _text(fields.get("type", MISSING)),
                             "filesystem": _text(fields.get("fs_type", MISSING)), "disabled": disabled, "coversInstallPath": covers})
    return _record_list(value, normalize, page)


def _volumes(value, page):
    def normalize(name, raw):
        fields = consumed_fields(raw, ("mount", "persistent", "disabled"))
        return omit_missing({"name": name, "mount": _text(fields.get("mount", MISSING)),
                             "persistent": _flag(fields.get("persistent", MISSING)), "disabled": _flag(fields.get("disabled", MISSING))})
    return _record_list(value, normalize, page)


def _attachment(value):
    if not isinstance(value, str):
        return unavailable("No simple additional-spec parameter reference is available; inspect the cluster template before attaching.")
    match = re.fullmatch(r"\$(?:([A-Za-z_][A-Za-z0-9_.-]{0,255})|\{([A-Za-z_][A-Za-z0-9_.-]{0,255})\})", value)
    if match is None:
        return unavailable("The attachment reference is not a simple parameter; inspect the cluster template before attaching.")
    return {"available": True, "parameterName": match.group(1) or match.group(2)}


def _identity(fields):
    name = required_wire_string(fields.get("name", MISSING), 256)
    array = _flag(fields.get("isarray", MISSING))
    if array is not True and _text(fields.get("template", MISSING)) != name:
        invalid_response()
    return {"name": name, "kind": "nodearray" if array is True else "node"}


def normalize_application_overview(value):
    result = []
    for raw in required_rows(value):
        fields = consumed_fields(raw, ("name", "template", "isarray", "state", "imagename", "slurmrole",
                                       "slurmpartition", "slurmhaenabled", "attachmentreference"))
        identity = _identity(fields)
        attachment = _attachment(fields.get("attachmentreference", MISSING))
        result.append(omit_missing({**identity, "state": _text(fields.get("state", MISSING)),
                                    "image": _identifier(fields.get("imagename", MISSING)),
                                    "role": _text(fields.get("slurmrole", MISSING)),
                                    "partition": _text(fields.get("slurmpartition", MISSING)),
                                    "haEnabled": _flag(fields.get("slurmhaenabled", MISSING)),
                                    "attachmentParameter": attachment.get("parameterName", MISSING)}))
    return _unique(result, lambda row: row["name"])


def normalize_application_details(value, install_path, section, page=None):
    section_fields = {
        "environment": ("extends", "state", "targetstate", "imagename", "machinetype", "architecture", "locker",
                        "slurmrole", "slurmversion", "slurmpartition", "slurmhaenabled", "slurmprimaryscheduler", "slurmautoscale"),
        "storage": ("mounts", "volumes"),
        "attachments": ("clusterinitspecs", "attachmentreference"),
    }
    if not isinstance(section, str) or section not in section_fields:
        invalid_response()
    if not is_install_path(install_path):
        invalid_response()
    page = {} if page is None else required_record(page)
    result = []
    for raw in required_rows(value):
        fields = consumed_fields(raw, ("name", "template", "isarray") + section_fields[section])
        identity = _identity(fields)
        if section == "storage":
            result.append({**identity, "mounts": _mounts(fields.get("mounts", MISSING), install_path, page),
                           "volumes": _volumes(fields.get("volumes", MISSING), page)})
        elif section == "attachments":
            result.append({**identity, "specs": _specs(fields.get("clusterinitspecs", MISSING), page),
                           "attachment": _attachment(fields.get("attachmentreference", MISSING))})
        else:
            image = _identifier(fields.get("imagename", MISSING))
            result.append(omit_missing({
                **identity, "state": _text(fields.get("state", MISSING)), "targetState": _text(fields.get("targetstate", MISSING)),
                "bases": _string_list(fields.get("extends", MISSING), page), "image": image,
                "platform": unavailable("The configured image was not supplied; its OS release is unresolved." if image is MISSING
                                        else "Image platform metadata could not be retrieved or validated."),
                "machineTypes": _string_list(fields.get("machinetype", MISSING), page),
                "architecture": _text(fields.get("architecture", MISSING)), "locker": _identifier(fields.get("locker", MISSING)),
                "scheduler": omit_missing({"role": _text(fields.get("slurmrole", MISSING)),
                                           "version": _text(fields.get("slurmversion", MISSING)),
                                           "partition": _text(fields.get("slurmpartition", MISSING)),
                                           "haEnabled": _flag(fields.get("slurmhaenabled", MISSING)),
                                           "primary": _flag(fields.get("slurmprimaryscheduler", MISSING)),
                                           "autoscale": _flag(fields.get("slurmautoscale", MISSING))}),
            }))
    return _unique(result, lambda row: row["name"])


def normalize_application_parameters(value, page=None):
    page = {} if page is None else required_record(page)
    result = []
    for raw in required_rows(value):
        fields = consumed_fields(raw, ("name", "label", "parametertype", "value"))
        if fields.get("parametertype") != "Cloud.ClusterInitSpecs":
            invalid_response()
        result.append(omit_missing({"name": required_wire_string(fields.get("name", MISSING), 256),
                                    "label": _text(fields.get("label", MISSING)), "specs": _specs(fields.get("value", MISSING), page)}))
    return _unique(result, lambda row: row["name"])


def normalize_context_cluster(value, name):
    items = required_rows(value)
    if not items:
        cluster_not_found()
    if len(items) != 1:
        invalid_response()
    fields = consumed_fields(items[0], ("clustername", "parentname", "state", "targetstate"))
    if required_wire_string(fields.get("clustername", MISSING), 256) != name:
        invalid_response()
    return omit_missing({"name": name, "parentName": _text(fields.get("parentname", MISSING)),
                         "state": _text(fields.get("state", MISSING)), "targetState": _text(fields.get("targetstate", MISSING))})
